distribution/lib/Standard/Base/0.0.0-dev/src/Data/Json/Extensions.enso

```diff
@@ -1,136 +1,137 @@
+   For arrays or vectors, the elements are converted recursively.
+   For atoms, the object is converted to a JS_Object.
+Vector.to_js_object : Vector
+Vector.to_js_object self =
+    proxy = Array_Proxy.new self.length i-> self.at i . to_js_object
+    Vector.from_polyglot_array proxy
+
+## Converts the given value to a JSON serializable object.
+   For Nothing, booleans, numbers and strings, this is the value itself.
+   For arrays or vectors, the elements are converted recursively.
+   For atoms, the object is converted to a JS_Object.
+Any.to_js_object : JS_Object
+Any.to_js_object self =
     m = Meta.meta self
     case m of
         _ : Meta.Atom ->
+            type_pair = ["type", Meta.get_qualified_type_name self . split '.' . last]
             cons = m.constructor
-            fs = m.fields
-            fnames = cons.fields
-            json_fs = 0.up_to fnames.length . fold Map.empty m-> i->
-                m.insert (fnames.at i) (fs.at i . to_json)
-            with_tp = json_fs . insert "type" (Json.String cons.name)
-            Json.Object with_tp
+            java_cons = cons.value ...
+            if java_cons == self then JS_Object.from_pairs [type_pair] else
+                fs = m.fields
+                field_names = cons.fields
+
+                builder = Vector.new_builder field_names.length+2
+                builder.append type_pair
+                builder.append ["constructor", cons.name]
+                0.up_to field_names.length . map i-> builder.append [field_names.at i, fs.at i . to_js_object]
+                JS_Object.from_pairs builder.to_vector
         _ : Meta.Constructor ->
-            Json.Object (Map.empty . insert "type" (Json.String m.name))
-
-        ## The following two cases cannot be handled generically and should
-           instead define their own `to_json` implementations.
-        _ : Meta.Polyglot -> Json.Null
-        _ : Meta.Primitive -> Json.Null
-
-## UNSTABLE
-
-   Returns a JSON representation of the dataflow error.
-
-   > Example
-     Converting a dataflow error to JSON.
-
-         import Standard.Examples
-
-         example_to_json = Examples.throw_error.to_json
-Error.to_json : Json
-Error.to_json self =
-    error_type = ["type", "Error"]
+            type_name = Meta.get_qualified_type_name self . split '.' . at -2
+            JS_Object.from_pairs [["type", type_name], ["constructor", m.name]]
+        _ -> Error.throw ("Cannot convert " + self.to_text + " to JSON")
+
+## Converts the given value to a JSON serializable object.
+   For Nothing, booleans, numbers and strings, this is the value itself.
+   For arrays or vectors, the elements are converted recursively.
+   For atoms, the object is converted to a JS_Object.
+Error.to_js_object : JS_Object
+Error.to_js_object self =
     caught = self.catch
-    error_content = ["content", caught.to_json]
+    error_type = ["type", "Error"]
+    error_content = ["content", caught]
```

Comment:
This logic seems to be copy paste of `Error.to_json`. Seems wrong.
Can't we keep it here and change `Error.to_json` to be `Json.stringify self.to_json_object`?